src/sys/select.rs

```diff
@@ -1,85 +1,75 @@
-use std::ptr::null_mut;
+use std::ptr::{null, null_mut};
 use std::os::unix::io::RawFd;
-use libc::c_int;
+use std::mem;
+use libc;
+use libc::{fd_set, c_int, timeval};
 use {Errno, Result};
 use sys::time::TimeVal;
+use sys::signal::SigSet;
 
-pub const FD_SETSIZE: RawFd = 1024;
-
-#[cfg(any(target_os = "macos", target_os = "ios"))]
-#[repr(C)]
 pub struct FdSet {
-    bits: [i32; FD_SETSIZE as usize / 32]
+    set: fd_set
 }
 
-#[cfg(any(target_os = "macos", target_os = "ios"))]
-const BITS: usize = 32;
-
-#[cfg(not(any(target_os = "macos", target_os = "ios")))]
-#[repr(C)]
-#[derive(Clone)]
-pub struct FdSet {
-    bits: [u64; FD_SETSIZE as usize / 64]
+impl AsRef<fd_set> for FdSet {
+    fn as_ref(&self) -> &fd_set {
+        &self.set
+    }
 }
 
-#[cfg(not(any(target_os = "macos", target_os = "ios")))]
-const BITS: usize = 64;
+pub const FD_SETSIZE: RawFd = libc::FD_SETSIZE as RawFd;
 
 impl FdSet {
     pub fn new() -> FdSet {
-        FdSet {
-            bits: [0; FD_SETSIZE as usize / BITS]
-        }
+        let mut set = FdSet {
+            set: unsafe { mem::uninitialized() }
+        };
+        set.clear();
+        set
     }
 
     pub fn insert(&mut self, fd: RawFd) {
-        let fd = fd as usize;
-        self.bits[fd / BITS] |= 1 << (fd % BITS);
+        assert!(fd >= 0 && fd < FD_SETSIZE, "RawFd out of bounds");
+        unsafe {
+            libc::FD_SET(fd, &mut self.set as *mut _);
+        }
     }
 
     pub fn remove(&mut self, fd: RawFd) {
-        let fd = fd as usize;
-        self.bits[fd / BITS] &= !(1 << (fd % BITS));
+        assert!(fd >= 0 && fd < FD_SETSIZE, "RawFd out of bounds");
+        unsafe {
+            libc::FD_CLR(fd, &mut self.set as *mut _);
+        }
     }
 
-    pub fn contains(&mut self, fd: RawFd) -> bool {
-        let fd = fd as usize;
-        self.bits[fd / BITS] & (1 << (fd % BITS)) > 0
+    pub fn contains(&self, fd: RawFd) -> bool {
+        assert!(fd >= 0 && fd < FD_SETSIZE, "RawFd out of bounds");
```

Comment:
we can probably just return false in this case, while preserving expected meaning?